rdr_service/resource/calculators/participant_enrollment_status.py

```diff
@@ -0,0 +1,303 @@
+#
+# This file is subject to the terms and conditions defined in the
+# file 'LICENSE', which is part of this source code package.
+#
+import datetime
+
+from rdr_service.code_constants import (
+    CONSENT_PERMISSION_YES_CODE,
+    CONSENT_PERMISSION_NO_CODE,
+    DVEHRSHARING_CONSENT_CODE_YES,
+    EHR_CONSENT_EXPIRED_YES,
+    DVEHRSHARING_CONSENT_CODE_NO, CONSENT_GROR_NO_CODE, CONSENT_GROR_NOT_SURE, CONSENT_GROR_YES_CODE
+)
+from rdr_service.resource.constants import PDREnrollmentStatusEnum
+from rdr_service.resource.constants import ParticipantEventEnum, COHORT_1_CUTOFF, \
+    COHORT_2_CUTOFF, ConsentCohortEnum
+from rdr_service.services.system_utils import JSONObject
+
+
+class EnrollmentStatusInfo:
+    """ Information about Enrollment Status events """
+    calculated = False  # False = No, True = Yes
+    first_ts = None  # First timestamp seen.
+    last_ts = None  # Last timestamp seen.
+    values: list = list()  # List of related events.
+
+
+class EnrollmentStatusCalculator:
+    """
+    Calculate participant enrollment status.
+    """
+    status: PDREnrollmentStatusEnum = PDREnrollmentStatusEnum.Unset
+    # events = None  # List of EnrollmentStatusEvent objects.
+    activity = None  # List of activity created from Participant generator.
+
+    cohort = None
+
+    # First info object for each part used in calculating enrollment status.
+    _signup = None
+    _consented = None
+    _ehr_consented = None
+    _gror_consented = None
+    _biobank_samples = None
+    _physical_measurements = None
+    _modules = None
+
+    # Timestamps for when each status was achieved, these are set in the self.save_calc() method.
+    registered_ts = None
+    participant_ts = None
+    participant_plus_ehr_ts = None
+    core_participant_minus_pm_ts = None
+    core_participant_ts = None
+
+    def run(self, activity: list):
+        """
+        :param activity: A list of activity dictionary objects created by the ParticipantSummaryGenerator.
+        """
+        self.activity = sorted([JSONObject(r) for r in activity if r['timestamp']], key=lambda i: i.timestamp)
+        # Work through activity by slicing to determine current enrollment status.
+        # This method allows us to iterate once through the data and still catch participants
+        # that might have been considered a Core Participant at one point, but would not by
+        # looking at their current state.
+        for x in range(1, len(self.activity)+1):
+            events = self.activity[0:x]
```

Comment:
Thanks, I think this simplifies the code greatly from what we had before in the generator.

Reply:
I think there might be a way to speed this up. From what I'm understanding this iterates through subsets of participant activity events. So, for a rough example, if a participant has the following events: `Primary Consent, EHR, GROR, Biosamples`. The algorithm would start by looking just at the Primary Consent and figure out the status from that. Then it would look at the status after the Primary Consent and the EHR. Then it would consider the status after Primary Consent, EHR, and GROR.

Each time it analyzes a new subset of events, it calls all of the methods for calculating the components of status (`calc_signup`, `calc_consent`, `calc_ehr_consent`, ...). And each of those methods loops through all of the events in the subset to search for the ones that are of interest to that method.

I think this looping and searching could be avoided by stepping through each event at the root level once, maintaining the activity seen in the timeline so far on the instance of the calculator, and calculating the status after each event has been analyzed. And each method would only be called with the event that it's looking for, rather than having the methods look through the entire list. So the code in the run function would look something like this:
```
for event in self.activity:
    if event.event == ParticipantEventEnum.SignupTime:
        self.signed_up = self.calc_signup(event)
    ...
```

Reply:
That is how the old enrollment status calculation code basically worked.  The issue there and possibly with your suggestion is we might have to add back in the additional code logic to see if a participant had reached core status at some point in the past, but wouldn't be if you look at their current data.  By working through the events in order regardless of event type, that removes the need for handling edge cases like participants that reached core status but then revoked their EHR consent later.

Reply:
What I'm describing is a bit different than how the old code worked, since the old code only calculated the status after all activity events had been processed. What I'm suggesting essentially boils down to iterating through the list of events once, calling the just the calc method that is designed for a given event's type (that would take the for-loop out of the `calc_...` methods), and check the status after each event has been analyzed (like it's doing now when it checks the status after a list of events was analyzed). What I'm suggesting would have one for-loop that goes through the events one at a time and still has the aggregate information for the events that had been seen before. Where the current implementation starts from the beginning each time another event is added (recalculating the same state that was found in the last iteration).

This is what the top-level method could look like:
```
for event in self.activity:
    if event.event == ParticipantEventEnum.SignupTime:
        self.signed_up = self.calc_signup(event)
    elif ev.event == ParticipantEventEnum.ConsentPII:
        self.consented, cohort = self.calc_consent(ev)
    elif ev.event == ParticipantEventEnum.EHRConsentPII or ev.event == ParticipantEventEnum.DVEHRSharing:
        self.ehr_consented = self.calc_ehr_consent(ev)
    elif ev.event == ParticipantEventEnum.GROR:
        self.gror_consented = self.calc_gror_consent(ev)
    elif ev.event == ParticipantEventEnum.BiobankConfirmed and ev.dna_tests > 0:
        self.biobank_samples = self.calc_biobank_samples(ev)
    elif ev.event == ParticipantEventEnum.PhysicalMeasurements and ev.finalized is not None:
        self.physical_measurements = self.calc_physical_measurements(ev)
    elif ev.group_id == 40:
        self.modules = self.calc_baseline_modules(ev)
```

And each calc method would be simplified to only handle an event that is already known to be the correct type. This is what the `calc_signup` method could look like:
```
         info = EnrollmentStatusInfo()
         info.calculated = True
         info.first_ts = ev.last_ts = ev.timestamp
         info.add_value(ev)
         return self.save_calc('_signup', info)
```

This would still catch any participants that have had CORE but don't because the calculation would still happen after each event and `self.status` wouldn't be downgraded if it was one of the CORE statuses.

Reply:
I'm not 100% sure this doesn't introduce opennings for edge cases to calculate wrong (I'll think more deeply about this on Monday).  The most obvious edge case is when there are many EHRConsent submissions with a No in the middle.  If it can be proven that it doesn't calculate incorrectly with edge case data, then I would be Ok with this change.

Reply:
What I would like to avoid in this code is having to add new code specifically to cover edge cases.  That made the older code pretty ugly to debug.

Reply:
My take:  I think we could look for performance improvements based on iterating over the chronologically sorted activity list once.  I think of this like a state machine, where the combination of the current enrollment status and the next event in the activity list determine the next state, and only certain events have an impact on certain states.  E.g., if you're CORE_MINUS_PM, you only care if you see a physical measurements event and don't have to process anything else.  If you reach CORE, you've reached the "terminal" state.

It would probably help to draw a visual diagram of the states and possible transitions based on tracked activities.

I think this could be considered for a future PR and does not have to be included now.   Changing the logic later shouldn't have to trigger any data rebuilds since it's just a different way to get to the same result.

Reply:
I agree with Peggy, I think the current code is in a good place where we can move forward and in the future we can iteratively improve it.